File: python/examples/trend_breakout.py
"""WMA 추세선 돌파 전략 (롱 온리, 다중 종목).

가중 이동 추세선으로 예측 고점/저점을 구하고, 현재가가 돌파선(예측고점+갭)을
넘고 상승 추세면 시장가 매수. 익절 +profit_rate, 손절 지지선(브라켓 위임).

실행:  HERMETIX_BROKER=next NEXT_CLIENT_ID=... python examples/trend_breakout.py
"""
from __future__ import annotations

import sys
from dataclasses import dataclass
from datetime import datetime, timedelta
from decimal import Decimal, ROUND_DOWN, ROUND_HALF_EVEN

sys.path.insert(0, __file__.rsplit("/", 2)[0])

from hermetix import Buy, Candle, CandleInterval, Sell, Signal, Strategy, StrategyContext, StrategySpec


@dataclass(frozen=True)
class TrendLine:
    high_price: Decimal
    low_price: Decimal
    high_inclination: Decimal
    low_inclination: Decimal
    avg_volume: Decimal

    @property
    def gap(self) -> Decimal:
        return self.high_price - self.low_price

    @property
    def breakout_line(self) -> Decimal:
        return self.high_price + self.gap

    @property
    def support_line(self) -> Decimal:
        return self.low_price - self.gap

    @staticmethod
    def of(candles: list[Candle]) -> "TrendLine":
        weight = Decimal(0)
        volume_weight = Decimal(1)
        high_inc = Decimal(0)
        low_inc = Decimal(0)
        avg_volume = Decimal(candles[0].volume)
        for i in range(1, len(candles)):
            w = Decimal(i)
            weight += w
            volume_weight += Decimal(i + 1)
            high_inc += (candles[i].high - candles[i - 1].high) * w
            low_inc += (candles[i].low - candles[i - 1].low) * w
            avg_volume += Decimal(candles[i].volume) * Decimal(i + 1)
        high_inc = (high_inc / weight).quantize(Decimal("0.00000001"), ROUND_HALF_EVEN)
        low_inc = (low_inc / weight).quantize(Decimal("0.00000001"), ROUND_HALF_EVEN)
        return TrendLine(
            high_price=candles[-1].high + high_inc,
            low_price=candles[-1].low + low_inc,
            high_inclination=high_inc,
            low_inclination=low_inc,
            avg_volume=(avg_volume / volume_weight).quantize(Decimal("0.00000001"), ROUND_HALF_EVEN),
        )


class TrendBreakoutStrategy(Strategy):

    def __init__(self, symbols: list[str] | None = None, lookback: int = 120,
                 profit_rate: Decimal = Decimal("0.04"), volume_ratio: Decimal = Decimal(0),
                 expire_hours: int = 12, budget_ratio: Decimal = Decimal("0.5"),
                 candle_interval: CandleInterval = CandleInterval.HOUR_1):
        self.symbols = symbols or ["AAPL"]
        self.lookback = lookback
        self.profit_rate = profit_rate
        self.volume_ratio = volume_ratio
        self.expire_hours = expire_hours
        self.budget_ratio = budget_ratio
        self.spec = StrategySpec(
            name="trend-breakout", symbols=self.symbols,
            candle_interval=candle_interval, candle_limit=lookback + 1,
        )
        self._last_entry_candle: dict[str, datetime] = {}
        self._entry_at: dict[str, datetime] = {}

    def decide(self, ctx: StrategyContext) -> list[Signal]:
        signals: list[Signal] = []
        for symbol in self.symbols:
            signals.extend(self._decide_symbol(symbol, ctx))
        return signals

    def _decide_symbol(self, symbol: str, ctx: StrategyContext) -> list[Signal]:
        if ctx.has_position(symbol):
            return self._decide_exit(symbol, ctx)
        self._entry_at.pop(symbol, None)
        if ctx.has_open_order(symbol):
            return []
        return self._decide_entry(symbol, ctx)

    def _decide_entry(self, symbol: str, ctx: StrategyContext) -> list[Signal]:
        candles = ctx.candles_of(symbol)
        if len(candles) < self.lookback + 1:
            return []
        completed, in_progress = candles[:-1], candles[-1]
        if self._last_entry_candle.get(symbol) == in_progress.timestamp:
            return []                       # 같은 시간봉 구간 재진입 방지

        trend = TrendLine.of(completed[-self.lookback:])
        if trend.high_inclination <= 0:     # 상승 추세만
            return []
        quote = ctx.quote(symbol)
        if quote is None or quote.price < trend.breakout_line:
            return []
        if self.volume_ratio > 0 and Decimal(in_progress.volume) < trend.avg_volume * self.volume_ratio:
            return []

        budget = ctx.buying_power * self.budget_ratio / len(self.symbols)
        quantity = (budget / quote.price).quantize(Decimal(1), rounding=ROUND_DOWN)
        if quantity < 1:
            return []

        self._last_entry_candle[symbol] = in_progress.timestamp
        self._entry_at[symbol] = ctx.now
        return [Buy(
            symbol=symbol, quantity=quantity,
            take_profit_price=(quote.price * (1 + self.profit_rate)).quantize(Decimal("0.01"), ROUND_HALF_EVEN),
            stop_loss_price=trend.support_line.quantize(Decimal("0.01"), ROUND_HALF_EVEN),
        )]

    def _decide_exit(self, symbol: str, ctx: StrategyContext) -> list[Signal]:
        opened_at = self._entry_at.setdefault(symbol, ctx.now)
        if ctx.now - opened_at < timedelta(hours=self.expire_hours):
            return []
        holding = ctx.holding(symbol)
        if holding is None:
            return []
        self._entry_at.pop(symbol, None)
        return [Sell(symbol=symbol, quantity=holding.quantity)]


if __name__ == "__main__":
    from _runner import run
    run(TrendBreakoutStrategy())

File: js/examples/trendBreakout.ts
/**
 * WMA 추세선 돌파 전략 (롱 온리, 다중 종목).
 * 예측고점+갭 돌파 & 상승 추세 -> 시장가 매수 (익절 +profitRate, 손절 지지선 - 브라켓 위임).
 *
 * 실행: HERMETIX_BROKER=next NEXT_CLIENT_ID=... node dist/examples/trendBreakout.js
 */
import { Decimal } from "decimal.js";
import type { Candle, CandleInterval, Signal, Strategy, StrategyContext, StrategySpec } from "../src/index.js";

export class TrendLine {
  constructor(
    readonly highPrice: Decimal,
    readonly lowPrice: Decimal,
    readonly highInclination: Decimal,
    readonly lowInclination: Decimal,
    readonly avgVolume: Decimal,
  ) {}

  get gap(): Decimal { return this.highPrice.minus(this.lowPrice); }
  get breakoutLine(): Decimal { return this.highPrice.plus(this.gap); }
  get supportLine(): Decimal { return this.lowPrice.minus(this.gap); }

  static of(candles: Candle[]): TrendLine {
    let weight = new Decimal(0);
    let volumeWeight = new Decimal(1);
    let highInc = new Decimal(0);
    let lowInc = new Decimal(0);
    let avgVolume = new Decimal(candles[0].volume);
    for (let i = 1; i < candles.length; i++) {
      weight = weight.plus(i);
      volumeWeight = volumeWeight.plus(i + 1);
      highInc = highInc.plus(candles[i].high.minus(candles[i - 1].high).mul(i));
      lowInc = lowInc.plus(candles[i].low.minus(candles[i - 1].low).mul(i));
      avgVolume = avgVolume.plus(new Decimal(candles[i].volume).mul(i + 1));
    }
    highInc = highInc.div(weight);
    lowInc = lowInc.div(weight);
    return new TrendLine(
      candles[candles.length - 1].high.plus(highInc),
      candles[candles.length - 1].low.plus(lowInc),
      highInc, lowInc, avgVolume.div(volumeWeight),
    );
  }
}

export interface TrendBreakoutOptions {
  symbols?: string[];
  lookback?: number;
  profitRate?: Decimal;
  volumeRatio?: Decimal;
  expireHours?: number;
  budgetRatio?: Decimal;
  candleInterval?: CandleInterval;
}

export class TrendBreakoutStrategy implements Strategy {
  readonly spec: StrategySpec;
  private readonly lookback: number;
  private readonly profitRate: Decimal;
  private readonly volumeRatio: Decimal;
  private readonly expireHours: number;
  private readonly budgetRatio: Decimal;
  readonly lastEntryCandle = new Map<string, number>();
  readonly entryAt = new Map<string, number>();

  constructor(opts: TrendBreakoutOptions = {}) {
    this.lookback = opts.lookback ?? 120;
    this.profitRate = opts.profitRate ?? new Decimal("0.04");
    this.volumeRatio = opts.volumeRatio ?? new Decimal(0);
    this.expireHours = opts.expireHours ?? 12;
    this.budgetRatio = opts.budgetRatio ?? new Decimal("0.5");
    this.spec = {
      name: "trend-breakout",
      symbols: opts.symbols ?? ["AAPL"],
      candleInterval: opts.candleInterval ?? "1h",
      candleLimit: this.lookback + 1,
    };
  }

  decide(ctx: StrategyContext): Signal[] {
    return this.spec.symbols.flatMap((symbol) => this.decideSymbol(symbol, ctx));
  }

  private decideSymbol(symbol: string, ctx: StrategyContext): Signal[] {
    if (ctx.hasPosition(symbol)) return this.decideExit(symbol, ctx);
    this.entryAt.delete(symbol);
    if (ctx.hasOpenOrder(symbol)) return [];
    return this.decideEntry(symbol, ctx);
  }

  private decideEntry(symbol: string, ctx: StrategyContext): Signal[] {
    const candles = ctx.candlesOf(symbol);
    if (candles.length < this.lookback + 1) return [];
    const completed = candles.slice(0, -1);
    const inProgress = candles[candles.length - 1];
    if (this.lastEntryCandle.get(symbol) === inProgress.timestamp.getTime()) return [];

    const trend = TrendLine.of(completed.slice(-this.lookback));
    if (trend.highInclination.lte(0)) return []; // 상승 추세만
    const quote = ctx.quote(symbol);
    if (!quote || quote.price.lt(trend.breakoutLine)) return [];
    if (this.volumeRatio.gt(0) && new Decimal(inProgress.volume).lt(trend.avgVolume.mul(this.volumeRatio))) return [];

    const budget = ctx.buyingPower.mul(this.budgetRatio).div(this.spec.symbols.length);
    const quantity = budget.div(quote.price).floor();
    if (quantity.lt(1)) return [];

    this.lastEntryCandle.set(symbol, inProgress.timestamp.getTime());
    this.entryAt.set(symbol, ctx.now.getTime());
    return [{
      kind: "buy", symbol, quantity,
      takeProfitPrice: quote.price.mul(new Decimal(1).plus(this.profitRate)).toDecimalPlaces(2),
      stopLossPrice: trend.supportLine.toDecimalPlaces(2),
    }];
  }

  private decideExit(symbol: string, ctx: StrategyContext): Signal[] {
    const openedAt = this.entryAt.get(symbol) ?? ctx.now.getTime();
    this.entryAt.set(symbol, openedAt);
    if (ctx.now.getTime() - openedAt < this.expireHours * 3600_000) return [];
    const holding = ctx.holding(symbol);
    if (!holding) return [];
    this.entryAt.delete(symbol);
    return [{ kind: "sell", symbol, quantity: holding.quantity }];
  }
}

if (import.meta.url === `file://${process.argv[1]}`) {
  const { run } = await import("./_runner.js");
  await run(new TrendBreakoutStrategy());
}
